pkg/spoke/controllers/statuscontroller/availablestatus_controller.go

```diff
@@ -0,0 +1,260 @@
+func aggregateManifestConditions(manifests []workapiv1.ManifestCondition) workapiv1.StatusCondition {
+	available, unavailable, unknown := 0, 0, 0
+	for _, manifest := range manifests {
+		for _, condition := range manifest.Conditions {
+			if condition.Type != string(workapiv1.ManifestAvailable) {
+				continue
+			}
+
+			switch condition.Status {
+			case metav1.ConditionTrue:
+				available += 1
+			case metav1.ConditionFalse:
+				unavailable += 1
+			case metav1.ConditionUnknown:
+				unknown += 1
+			}
+		}
+	}
+
+	switch {
+	case unavailable > 0:
+		return workapiv1.StatusCondition{
+			Type:    string(workapiv1.WorkAvailable),
+			Status:  metav1.ConditionFalse,
+			Reason:  "ResourcesNotAvailable",
+			Message: fmt.Sprintf("%d of %d resources are not available", unavailable, len(manifests)),
+		}
+	case unknown > 0:
+		return workapiv1.StatusCondition{
+			Type:    string(workapiv1.WorkAvailable),
+			Status:  metav1.ConditionUnknown,
+			Reason:  "ResourcesStatusUnknown",
+			Message: fmt.Sprintf("%d of %d resources have unknown status", unknown, len(manifests)),
+		}
+	default:
+		return workapiv1.StatusCondition{
+			Type:    string(workapiv1.WorkAvailable),
+			Status:  metav1.ConditionTrue,
+			Reason:  "ResourcesAvailable",
+			Message: "All resources are available",
+		}
+	}
+}
+
+// buildAvailableStatusCondition returns a StatusCondition with type Available for a given manifest resource
+func buildAvailableStatusCondition(resourceMeta workapiv1.ManifestResourceMeta, dynamicClient dynamic.Interface) workapiv1.StatusCondition {
+	conditionType := string(workapiv1.ManifestAvailable)
+
+	if len(resourceMeta.Resource) == 0 || len(resourceMeta.Version) == 0 || len(resourceMeta.Name) == 0 {
+		return workapiv1.StatusCondition{
+			Type:    conditionType,
+			Status:  metav1.ConditionUnknown,
+			Reason:  "IncompletedResourceMeta",
+			Message: "Resource meta is incompleted",
+		}
+	}
+
+	available, err := isResourceAvailable(resourceMeta.Namespace, resourceMeta.Name, schema.GroupVersionResource{
+		Group:    resourceMeta.Group,
+		Version:  resourceMeta.Version,
+		Resource: resourceMeta.Resource,
+	}, dynamicClient)
+	if err != nil {
+		return workapiv1.StatusCondition{
+			Type:    conditionType,
+			Status:  metav1.ConditionUnknown,
+			Reason:  "FetchingResourceFailed",
+			Message: fmt.Sprintf("Failed to fetch resource: %v", err),
+		}
+	}
+
+	if available {
+		return workapiv1.StatusCondition{
+			Type:    conditionType,
+			Status:  metav1.ConditionTrue,
+			Reason:  "ResourceAvailable",
+			Message: "Resource is available",
+		}
+	}
+
+	return workapiv1.StatusCondition{
+		Type:    conditionType,
+		Status:  metav1.ConditionFalse,
+		Reason:  "ResourceNotAvailable",
+		Message: "Resource is not available",
+	}
+}
+
+// isResourceAvailable checks if the specific resource is available or not
+func isResourceAvailable(namespace, name string, gvr schema.GroupVersionResource, dynamicClient dynamic.Interface) (bool, error) {
```

Comment:
If a work has a resource X that has spec A, and then the work is updated to have spec B, is that resource still available?

Reply:
I'm having a hard time understanding how something becomes 'unavailable' until it is wholly removed from the work and the spoke.

Reply:
> If a work has a resource X that has spec A, and then the work is updated to have spec B, is that resource still available?

If the spec update does not change resource meta (including GVR, namespace and name), the resource will be thought available no matter spec B is applied successfully or not; otherwise it depends on whether spec B is applied successfully.

Reply:
> I'm having a hard time understanding how something becomes 'unavailable' until it is wholly removed from the work and the spoke.

If the controller cannot find the resource with resource meta (including GVR, namespace and name), which is extracted either from apply result of manifest or from manifest itself, the manifest will have an Available condition with status False. It happens when the manifest is failed to apply or the applied resource is deleted by others for some reason.